### PR TITLE
BUG: Fix F2py with meson when fortran code uses "include" statement

### LINKED ISSUE #25344: BUG: f2py (3.12) with meson does not support correct include file
### Describe the issue:

When running f2py with python3.12 (so using meson). I do face the following issue:
```
gfortran -Iuntitled.cpython-312-darwin.so.p -I. -I.. -I/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/numpy/core/include -I/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/numpy/f2py/src -I/Library/Frameworks/Python.framework/Versions/3.12/include/python3.12 -fvisibility=hidden -fdiagnostics-color=always -Wall -O3 -Juntitled.cpython-312-darwin.so.p -o untitled.cpython-312-darwin.so.p/matrix.f.o -c ../matrix.f
../matrix.f:324:0:

  324 |       include 'coupl.inc'
      |
Fatal Error: Cannot open included file ‘coupl.inc’
compilation terminated.
```

The reason is not the trivial one (I do have such file in the local directory) but the fact that such file is not copy to the meson build directory. Additionally any attempt to provide a path to a directory where such file exists fails 
(either by specifying --include-paths or -I or even the outdated --include_paths). To be precise the failing compilation line is not updated correctly by any of those options.

### Reproduce the code example:

```python
C
      SUBROUTINE DFILTER2D(A,B,M,N)
C
       INCLUDE 'AB.inc' # put the following line inside that file
C      DOUBLE PRECISION A(M,N)
C      DOUBLE PRECISION B(M,N)
      INTEGER N, M
CF2PY INTENT(OUT) :: B
CF2PY INTENT(HIDE) :: N
CF2PY INTENT(HIDE) :: M
      DO 20 I = 2,M-1
         DO 40 J = 2,N-1
            B(I,J) = A(I,J) +
     &           (A(I-1,J)+A(I+1,J) +
     &           A(I,J-1)+A(I,J+1) )*0.5D0 +
     &           (A(I-1,J-1) + A(I-1,J+1) +
     &           A(I+1,J-1) + A(I+1,J+1))*0.25D0
 40      CONTINUE
 20   CONTINUE
      END
```

### Error message:

```shell
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.12/bin/f2py", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/numpy/f2py/f2py2e.py", line 732, in main
    run_compile()
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/numpy/f2py/f2py2e.py", line 705, in run_compile
    builder.compile()
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/numpy/f2py/_backends/_meson.py", line 131, in compile
    self.run_meson(self.build_dir)
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/numpy/f2py/_backends/_meson.py", line 124, in run_meson
    raise subprocess.CalledProcessError(
subprocess.CalledProcessError: Command '['meson', 'compile', '-C', 'bbdir']' returned non-zero exit status 1.
```

### Python and NumPy Versions:

>>> import sys, numpy; print(numpy.__version__); print(sys.version)
1.26.2
3.12.0 (v3.12.0:0fb18b02c8, Oct  2 2023, 09:45:56) [Clang 13.0.0 (clang-1300.0.29.30)]

### Runtime Environment:

[{'numpy_version': '1.26.2',
  'python': '3.12.0 (v3.12.0:0fb18b02c8, Oct  2 2023, 09:45:56) [Clang 13.0.0 '
            '(clang-1300.0.29.30)]',
  'uname': uname_result(system='Darwin', node='mac-1q0-072.local', release='22.1.0', version='Darwin Kernel Version 22.1.0: Sun Oct  9 20:15:09 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T6000', machine='arm64')},
 {'simd_extensions': {'baseline': ['NEON', 'NEON_FP16', 'NEON_VFPV4', 'ASIMD'],
                      'found': ['ASIMDHP'],
                      'not_found': ['ASIMDFHM']}},
 {'architecture': 'armv8',
  'filepath': '/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/numpy/.dylibs/libopenblas64_.0.dylib',
  'internal_api': 'openblas',
  'num_threads': 10,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.23.dev'}]
None

### Context for the issue:

This is a blocker for supporting python3.12 for the moment so not directly critical.

### PR DESCRIPTION
**Pull request by oliviermattelaer**
This is a proposal to fix #25344 

In practice, I'm adding the current work directory to the list of  include paths.
Obviously, they are likely a nicer way to fix the issue. So do not hesistate to trash it to have something nicer/more robust/...

Thanks,

Olivier

### CODE REVIEW

File: numpy/f2py/_backends/_meson.py [hunk 1]
```diff
@@ -59,6 +59,7 @@ def meson_build_template(self) -> str:
     def initialize_template(self) -> None:
         self.substitutions["modulename"] = self.modulename
         self.substitutions["buildtype"] = self.build_type
+        self.substitutions["original_source_dir"] = os.getcwd()
```
[the remark this answers is not in the document]
Comment (oliviermattelaer) [reply numpy/f2py/_backends/_meson.py:62]: This resolves to the directory where f2py is run.

[the remark this answers is not in the document]
Comment (HaoZeke) [reply numpy/f2py/_backends/_meson.py:62]: That would be incorrect behavior in general. I believe the fix should be to make sure all the relevant input files are copied into the directory where `meson` executes, and then that path should be added to `-I`.

[the remark this answers is not in the document]
Comment (oliviermattelaer) [reply numpy/f2py/_backends/_meson.py:62]: So you want to parse the fortran file for searching the include statement and then move the corresponding file?
Or you want to include all the file that ends with ".inc"?

**Comment by HaoZeke** [comment]
I think the pragmatic solution for now is to move all `.inc` files.

File: numpy/f2py/_backends/_meson.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (HaoZeke) [reply numpy/f2py/_backends/_meson.py:62]: Yup, that would make sense for now. Either that or move all files where `f2py` is present, and add the build directory as a `-I`

[the remark this answers is not in the document]
Comment (HaoZeke) [reply numpy/f2py/_backends/_meson.py:62]: Thinking about this a bit more, I realize that we can't just copy all the files over into a temporary directory. The user should be explicit about the files they need (including `inc` files).

@oliviermattelaer under `distutils` does `f2py` effectively pass `-I {directory_f2py_was_run}`? That seems very confusing in practice (e.g. if there's an existing shared library. We can (and if it is the older behavior) bring this back, but with a note in the documentation.

**Comment by HaoZeke** [comment]
Closing as superseeded by https://github.com/numpy/numpy/pull/25364

**Pull request closed without merging**